rfcs/text/0015_ols_phase_1.md

```diff
@@ -0,0 +1,268 @@
+
+```ts
+/**
+ * The accessClassification dictates the protection level of the saved object:
+ *  * public (default): instances of this saved object type will be accessible to all users within the given namespace, who are authorized to act on objects of this type.
+ *  * private: instances of this saved object type will belong to the user who created them, and will not be accessible by other users, except for administrators.
+ */
+export type SavedObjectsAccessClassification = 'public' | 'private';
+
+// Note: some existing properties have been omitted for brevity.
+export interface SavedObjectsType {
+  name: string;
+  hidden: boolean;
+  namespaceType: SavedObjectsNamespaceType;
+  mappings: SavedObjectsTypeMappingDefinition;
+
+  /**
+   * The {@link SavedObjectsAccessClassification | accessClassification} for the type.
+   */
+  accessClassification?: SavedObjectsAccessClassification;
+}
+
+// Example consumer
+class MyPlugin {
+  setup(core: CoreSetup) {
+    core.savedObjects.registerType({
+      name: 'user-settings',
+      accessClassification: 'private',
+      namespaceType: 'single',
+      hidden: false,
+      mappings,
+    })
+  }
+}
+```
+
+### 3.1.2 Schema
+Saved object ownership will be recorded as metadata within each `private` saved object. We do so by adding a top-level `accessControl` object with a singular `owner` property:
+
+```ts
+/**
+ * Describes which users should be authorized to access this SavedObject.
+ *
+ * @public
+ */
+export interface SavedObjectAccessControl {
+  /** The owner of this SavedObject. */
+  owner: string;
+}
+
+// Note: some existing fields have been omitted for brevity
+export interface SavedObject<T = unknown> {
+  id: string;
+  type: string;
+  attributes: T;
+  references: SavedObjectReference[];
+  namespaces?: string[];
+  /** Describes which users should be authorized to access this SavedObject. */
+  accessControl?: SavedObjectAccessControl;
+}
+```
+
+### 3.1.3 Saved Objects Client: Security wrapper
+
+The [security wrapper](https://github.com/elastic/kibana/blob/701697cc4a34d07c0508c3bdf01dca6f9d40a636/x-pack/plugins/security/server/saved_objects/secure_saved_objects_client_wrapper.ts) authorizes and audits operations against saved objects.
+
+There are two primary changes to this wrapper:
+
+#### Attaching Access Controls
+
+This wrapper will be responsible for attaching an access control specification to all private objects before they are created.
+It will also allow users to provide their own access control specification in order to support the import/create use cases.
+
+Similar to the way we treat `namespaces`, it will not be possible to change an access control specification via the `update`/`bulk_update` functions in this first phase. We may consider adding a dedicated function to update the access control specification, similar to what we've done for sharing to spaces.
```

Comment:
Why should changing the ACL / owner be a separate API? Could we just prevent changing owners from the client but allow changing the acl from the repository?

This is very much forward-looking, but in a world where we've deprecated the saved objects client and plugins only consume the repository from behind their API controllers, it feels like it could simplify things if a plugin is able to just update a saved object and it's ACL in a single api method.

Reply:
I don't expect to need either of these mechanisms for this first phase of work, so this is something we can discuss in more detail when the time comes. My primary motivation for making this a separate API is audit logging. Auditors might want to know when an object has changed owners, or when a user has chosen to share their object with another user.

The normal `update` API doesn't record this level of granularity in the audit logs. We could add conditional logic to audit ACL updates differently from other updates, but my preference would be to treat this as a distinct operation.

Secondly, it's plausible that a user would be authorized to update a saved object, but unauthorized to adjust the ACL. Allowing the ACL to be updated via `update` would further complicate the authorization logic here.